### PR TITLE
fixed the last failing test in dom/nodes/attributes.html

### PR DESCRIPTION
**Pull request by ghost**
r? @nox

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9800)

### CODE REVIEW

File: components/script/dom/namednodemap.rs [hunk 1]
```diff
@@ -99,8 +99,21 @@ impl NamedNodeMapMethods for NamedNodeMap {
...
+                      if s != *previous {
+                          *previous = s.clone();
+                          if self.owner.html_element_in_html_document() && s != &*s.to_lowercase() {
+                              Some(None)
+                          } else {
+                              Some(Some(s))
+                          }
+                      } else {
+                          Some(None)
+                      }
```
Comment (Wafflespeanut) [inline components/script/dom/namednodemap.rs]: I'm not sure whether this is clean :/

File: components/script/dom/namednodemap.rs [hunk 2]
```diff
@@ -99,8 +99,15 @@ impl NamedNodeMapMethods for NamedNodeMap {
...
     // https://heycam.github.io/webidl/#dfn-supported-property-names
     fn SupportedPropertyNames(&self) -> Vec<DOMString> {
-        self.owner.attrs().iter().map(|attr| {
-            DOMString::from(&**attr.name()) // FIXME(ajeffrey): Convert directly from &Atom to DOMString
-        }).collect()
+        let mut string_vec = vec!();
+        for attr in self.owner.attrs().iter() {
+            // FIXME(ajeffrey): Convert directly from &Atom to DOMString
+            let s = DOMString::from(&**attr.name());
+            if self.owner.html_element_in_html_document() && s != &*s.to_lowercase() {
```
Comment (Ms2ger) [inline components/script/dom/namednodemap.rs]: `to_lowercase` is almost always wrong in DOM code; try to write a test with uppercase characters outside A-Z that fails with this code.

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/script/dom/namednodemap.rs, line 105 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9800#-KBeZtzPU8M8giuCKcza:-KBeZtzPU8M8giuCKczb:108790666) ([raw file](https://github.com/servo/servo/blob/26d167677e1cd6222b2dd7f5f4bb3d15e9cf01eb/components/script/dom/namednodemap.rs#L105)):*
You can remove this.

---

*[components/script/dom/namednodemap.rs, line 102 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9800#-KBgZi5MOw66bqR4vO6V:-KBgZi5MOw66bqR4vO6W:1108947753) ([raw file](https://github.com/servo/servo/blob/501c0abb6a3bc128fd0752a0ee4a541a0f14c052/components/script/dom/namednodemap.rs#L102)):*
Just name that `names`.

---

*[components/script/dom/namednodemap.rs, line 106 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9800#-KBgWMV9bizz6DYLMZ-u:-KBgWMV9bizz6DYLMZ-v:-392774569) ([raw file](https://github.com/servo/servo/blob/501c0abb6a3bc128fd0752a0ee4a541a0f14c052/components/script/dom/namednodemap.rs#L106)):*
Please bind `self.owner.html_element_in_html_document()` to a local variable outside the loop.

---

*[components/script/dom/namednodemap.rs, line 106 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9800#-KBeYSrdIa2k_1Kg0jp5-r2-106:-KBgWSr-4H-KpcLACGRg:382823212) ([raw file](https://github.com/servo/servo/blob/501c0abb6a3bc128fd0752a0ee4a541a0f14c052/components/script/dom/namednodemap.rs#L106)):*
Yes, and instead of actually converting to lowercase, you can do something like:

```rust
if s.bytes().all(|b| b.to_ascii_lowercase() == b) {
    continue;
}
```

And you should do this on the atom's text before converting it to `DOMString`.

---

*[components/script/dom/namednodemap.rs, line 111 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9800#-KBgZmekVZjMZEIh0aCf:-KBgZmekVZjMZEIh0aCg:-1781374906) ([raw file](https://github.com/servo/servo/blob/501c0abb6a3bc128fd0752a0ee4a541a0f14c052/components/script/dom/namednodemap.rs#L111)):*
Nit: missing new line.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9800#-:-KBg_0WuFUe9wPAjZ6R0:1458266033)*

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/script/dom/namednodemap.rs, line 108 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9800#-KC0juBtOEhAVLBureQt:-KC0juBuqWBPpEqPR41Q:-372090915) ([raw file](https://github.com/servo/servo/blob/fe2b585b2a516c3777b9cc1d629559284849ae39/components/script/dom/namednodemap.rs#L108)):*
This still allocates a new lowercased string for no good reason. Please use a loop to lowercase each char on their own instead, like I stated in my previous comment.

---

*[components/script/dom/namednodemap.rs, line 110 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9800#-KC0ktS9-Lq1vEwEWioH:-KC0ktS9-Lq1vEwEWioI:2099419735) ([raw file](https://github.com/servo/servo/blob/fe2b585b2a516c3777b9cc1d629559284849ae39/components/script/dom/namednodemap.rs#L110)):*
Nit: no need for `else` given the previous block continues execution.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9800#-:-KC0l-P7yDUbXWirla4o:1458266033)*

**Comment by Ms2ger** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/script/dom/namednodemap.rs, line 108 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9800#-KC0juBtOEhAVLBureQt:-KC0oiJNYwXFFfC0YRsC:545223237) ([raw file](https://github.com/servo/servo/blob/fe2b585b2a516c3777b9cc1d629559284849ae39/components/script/dom/namednodemap.rs#L108)):*
Not that `Atom::to_ascii_lowercase` doesn't allocate in the common case that the name is already lowercase.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9800)*

**Comment by nox** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/script/dom/namednodemap.rs, line 108 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9800#-KC0juBtOEhAVLBureQt:-KC0uYNFDm295brbhrxS:420221446) ([raw file](https://github.com/servo/servo/blob/fe2b585b2a516c3777b9cc1d629559284849ae39/components/script/dom/namednodemap.rs#L108)):*
Oh right, but it still bumps a refcount, and then we compare the lowercased version against the original one. Looping over the bytes is still better.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9800)*

**Comment by Ms2ger** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/script/dom/namednodemap.rs, line 108 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9800#-KC0juBtOEhAVLBureQt:-KC0xEyrH1d-ZP_xZTkA:-1292737794) ([raw file](https://github.com/servo/servo/blob/fe2b585b2a516c3777b9cc1d629559284849ae39/components/script/dom/namednodemap.rs#L108)):*
`eq_ignore_ascii_case`, then

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9800)*

**Comment by nox** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/script/dom/namednodemap.rs, line 108 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9800#-KC0juBtOEhAVLBureQt:-KC0yuyNOOcTb8QX9KC-:1386982831) ([raw file](https://github.com/servo/servo/blob/fe2b585b2a516c3777b9cc1d629559284849ae39/components/script/dom/namednodemap.rs#L108)):*
No. We have only the original version and don't want to do anything if it isn't lowercased. Looping over the bytes is still better.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9800)*

**Comment by nox** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/script/dom/namednodemap.rs, line 104 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9800#-KC14VDWRwnJNwoxqrow:-KC14VDWRwnJNwoxqrox:1778143627) ([raw file](https://github.com/servo/servo/blob/fe2b585b2a516c3777b9cc1d629559284849ae39/components/script/dom/namednodemap.rs#L104)):*
Nit: just call that `html_element_in_html_document`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9800)*

**Comment by Wafflespeanut** [comment]
@nox addressed the comments. r?

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[components/script/dom/namednodemap.rs, line 114 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9800#-KC5G5-NRghSkwJLIVh4:-KC5G5-NRghSkwJLIVh5:-778882556) ([raw file](https://github.com/servo/servo/blob/5d9444664e685f1cee8497741bcdf6036baaad88/components/script/dom/namednodemap.rs#L114)):*
Avoid a useless allocation by doing:

```rust
if !names.any(|name| &*name == s) {
    names.push(DOMString::from(s));
}
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9800#-:-KC5GpFPGiXYRj4kiDHl:1458266033)*

**Comment by nox** [comment]
@bors-servo r+

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9800#-:-KC5dlPcSpKeOks3Yz18:1236647940)*

**Comment by Manishearth** [comment]
And back to your regularly scheduled borsgramming.

**Pull request merged**